src/snowflake/telemetry/serialize/__init__.py

```diff
@@ -0,0 +1,172 @@
+import struct
+from enum import IntEnum
+from typing import Callable, Any, List, Union
+
+Enum = IntEnum
+
+class MessageMarshaler:
+    def __bytes__(self) -> bytes:
+        out = ProtoSerializer()
+        self.write_to(out)
+        return bytes(out)
+
+    def write_to(self, out: "ProtoSerializer") -> None:
+        ...
+
+class ProtoSerializer:
```

Comment:
One optimization that you can do is to pre-calculate the size of the binary array and pre-allocate the array.

Reply:
I tried changing my `serialize` implementation to write to a right-sized buffer (using memoryview for zero-copy fast overwriting of slices). I noticed that it was about 10-20% slower than my current implementation (Can see my changes here https://github.com/sfc-gh-jopel/snowflake-telemetry-python/pull/1/files). This script illustrates it a little. 
```
def test1():
    b = bytearray(len(b"hello world") * 10000)
    m = memoryview(b)
    i = 0
    l = len(b"hello world")
    for _ in range(10000):
        m[i:i+l] = b"hello world"
        i += l
    return b

def test2():
    b = bytearray()
    for _ in range(10000):
        b.extend(b"hello world"[::-1])
    return b[::-1]

def test3():
    b = bytearray()
    for _ in range(10000):
        b += b"hello world"[::-1]
    return b[::-1]

if __name__ == "__main__":
    import timeit
    print(timeit.timeit("test1()", setup="from __main__ import test1", number=1000))
    print(timeit.timeit("test2()", setup="from __main__ import test2", number=1000))
    print(timeit.timeit("test3()", setup="from __main__ import test3", number=1000))

```

Output:
```
1.3387652660021558
1.4347571920079645
1.276998276996892
```

I had the `extend` approach reverse the bytes before writing since that is what my current implementation is doing. The preallocated approach was only slightly faster than the `extend` approach, and slower than `+=` method. (I should switch the `extend` to `+=` for a slight performance increase). Calculating the size of the buffer would be more complex for our proto messages than what is done in this script, and likely make it the slowest of the 3. 

This result is puzzling to me because the `memoryview` approach should be more memory friendly. My guess is for the messages I am using in my benchmark (and probably most telemetry data) fits in cache so the resize operation happens infrequently and is fast enough that it still outperforms the overhead of bounds checking in the right-sized approach.

I am not sure what is the best thing to do here. I think `memoryview` will outperform for huge logs and traces, but I don't think it's the typical use case.